syncstorage-postgres/src/db/mod.rs

```diff
@@ -1,16 +1,358 @@
+    in_transaction: bool,
+    /// Boolean to identify if query in active transaction.
+    in_write_transaction: bool,
+    /// Whether update_collection has already been called.
+    updated_collection: bool,
+}
+
+impl PgDb {
+    /// Create a new instance of PgDb
+    /// Fresh metrics clone and default impl of session.
+    pub(super) fn new(
+        conn: Conn,
+        coll_cache: Arc<CollectionCache>,
+        metrics: &Metrics,
+        quota: &Quota,
+    ) -> Self {
+        PgDb {
+            conn,
+            session: Default::default(),
+            coll_cache,
+            metrics: metrics.clone(),
+            quota: *quota,
+        }
+    }
+
+    /// NOTE: Will be completed with other db method task.
+    pub(super) async fn get_or_create_collection_id(&mut self, _name: &str) -> DbResult<i32> {
+        todo!()
+    }
+
+    async fn commit(&mut self) -> DbResult<()> {
+        if self.session.in_transaction {
+            <Conn as AsyncConnection>::TransactionManager::commit_transaction(&mut self.conn)
+                .await?;
+        }
+        Ok(())
+    }
+
+    /// Utility to rollback transaction if current Db session transaction in progress.
+    async fn rollback(&mut self) -> DbResult<()> {
+        if self.session.in_transaction {
+            <Conn as AsyncConnection>::TransactionManager::rollback_transaction(&mut self.conn)
+                .await?;
+        }
+        Ok(())
+    }
+
+    /// Utility method to begin transaction and set current session `in_transaction` to `true`
+    /// If `for_write` truthy, `in_write_transaction` sets to true.
+    pub(super) async fn begin(&mut self, for_write: bool) -> DbResult<()> {
+        <Conn as AsyncConnection>::TransactionManager::begin_transaction(&mut self.conn).await?;
+        self.session.in_transaction = true;
+        if for_write {
+            self.session.in_write_transaction = true;
+        }
+        Ok(())
+    }
+
+    /// Simple check function to ensure database liveliness.
+    async fn check(&mut self) -> DbResult<results::Check> {
+        diesel::sql_query("SELECT 1")
+            .execute(&mut self.conn)
+            .await?;
+        Ok(true)
+    }
+
+    pub(super) fn timestamp(&self) -> SyncTimestamp {
+        self.session.timestamp
+    }
+}
+
+#[async_trait(?Send)]
+impl Db for PgDb {
+    async fn commit(&mut self) -> Result<(), Self::Error> {
+        PgDb::commit(self).await
+    }
+
+    async fn rollback(&mut self) -> Result<(), Self::Error> {
+        PgDb::rollback(self).await
+    }
+
+    async fn begin(&mut self, for_write: bool) -> Result<(), Self::Error> {
+        PgDb::begin(self, for_write).await
+    }
+
+    async fn check(&mut self) -> Result<results::Check, Self::Error> {
+        PgDb::check(self).await
+    }
+
+    async fn lock_for_read(
+        &mut self,
+        params: params::LockCollection,
+    ) -> Result<results::LockCollection, Self::Error> {
+        PgDb::lock_for_read(self, params).await
+    }
```

Comment:
Like what the spanner impl has, let's move this `impl Db` into a `db/db_impl.rs` and `impl BatchDb` into `db/batch_impl.rs`.

Also we don't need the older style, extra level of redirection here either: the actual impl of these methods should reside in these trait method definitions themselves, not in e.g. `PgDb::commit`. So the `commit/rollback/begin/check/timestamp` methods above in `PgDb` should move into here, and every other stub'd method should be a `todo!()` for now instead.

```suggestion
impl Db for PgDb {
    async fn commit(&mut self) -> Result<(), Self::Error> {
        <impl from above>
    }

    async fn rollback(&mut self) -> Result<(), Self::Error> {
        <impl from above>
    }

    async fn begin(&mut self, for_write: bool) -> Result<(), Self::Error> {
        <impl from above>
    }

    async fn check(&mut self) -> Result<results::Check, Self::Error> {
        <impl from above>
    }

    async fn lock_for_read(
        &mut self,
        params: params::LockCollection,
    ) -> Result<results::LockCollection, Self::Error> {
        todo!()
    }
```

Reply:
Should we merge this in first @pjenvey and fold into your changes in #1882 or other way around?

Reply:
Sure -- either way should work, I don't think these PRs touch any of the same files (I didn't touch syncstorage-postgres)